ding/envs/env_manager/envpool_env_manager.py

```diff
@@ -124,3 +156,158 @@ def env_num(self) -> int:
+
+    @property
+    def action_space(self) -> 'gym.spaces.Space':  # noqa
+        try:
+            return self._action_space
+        except AttributeError:
+            self.launch()
+            self.close()
+            return self._action_space
+
+
+@ENV_MANAGER_REGISTRY.register('env_pool_v4')
+class PoolEnvManagerV2:
+    '''
+    Overview:
+        PoolEnvManagerV2 supports new pipeline of DI-engine.
+        Envpool now supports Atari, Classic Control, Toy Text, ViZDoom.
+        Here we list some commonly used env_ids as follows.
+        For more examples, you can refer to <https://envpool.readthedocs.io/en/latest/api/atari.html>.
+
+        - Atari: "Pong-v5", "SpaceInvaders-v5", "Qbert-v5"
+        - Classic Control: "CartPole-v0", "CartPole-v1", "Pendulum-v1"
+    '''
+
+    @classmethod
+    def default_config(cls) -> EasyDict:
+        return EasyDict(deepcopy(cls.config))
+
+    config = dict(
+        type='envpool',
+        env_num=8,
+        batch_size=8,
+        image_observation=True,
+    )
+
+    def __init__(self, cfg: EasyDict) -> None:
+        super().__init__()
+        self._cfg = self.default_config()
+        self._cfg.update(cfg)
+        self._env_num = cfg.env_num
+        self._batch_size = cfg.batch_size
+
+        self._closed = True
+        self._seed = None
+        self._test = False
+
+    def launch(self) -> None:
+        assert self._closed, "Please first close the env manager"
+        if self._seed is None:
+            seed = 0
+        else:
+            seed = self._seed
+
+        kwargs = {}
+        if "episodic_life" in self._cfg:
+            kwargs["episodic_life"] = self._cfg.episodic_life
+        if "reward_clip" in self._cfg:
+            kwargs["reward_clip"] = self._cfg.reward_clip
+        if "stack_num" in self._cfg:
+            kwargs["stack_num"] = self._cfg.stack_num
+        if "gray_scale" in self._cfg:
+            kwargs["gray_scale"] = self._cfg.gray_scale
+        if "frame_skip" in self._cfg:
+            kwargs["frame_skip"] = self._cfg.frame_skip
+        if "test" in self._cfg:
+            self._test = self._cfg.test
+
+        self._envs = envpool.make(
+            task_id=self._cfg.env_id,
+            env_type="gym",
+            num_envs=self._env_num,
+            batch_size=self._batch_size,
+            seed=seed,
+            **kwargs
+        )
+        self._action_space = self._envs.action_space
+        self._observation_space = self._envs.observation_space
+        self._closed = False
+        return self.reset()
+
+    def reset(self) -> None:
+        self._envs.async_reset()
+        ready_obs = {}
+        while True:
+            obs, _, _, info = self._envs.recv()
+            env_id = info['env_id']
+            obs = obs.astype(np.float32)
+            if self._cfg.image_observation:
+                obs /= 255.0
+            ready_obs = deep_merge_dicts({i: o for i, o in zip(env_id, obs)}, ready_obs)
```

Comment:
directly use the assignment operation, don't use `deep_merge_dicts` too many times, it is a complex function

Reply:
fixed